### PR TITLE
cm/dashboard merge main

### PR DESCRIPTION
**Pull request by CharlieMcVicker**
This is PR #352 with main merged in and `dev-check` (formatting) run.

### CODE REVIEW

**Comment by GracefulLemming** [comment]
Make changes as mentioned here: https://github.com/NEU-DSG/dailp-encoding/pull/358#discussion_r1408431394

File: website/src/pages/documents/document.page.tsx [hunk 1]
```diff
@@ -480,3 +488,48 @@ export const DocumentTitleHeader = (p: {
     )}
   </header>
 )
+
+/** Button that allows users to bookmark a document */
+export const BookmarkButton = (props: { documentId: String }) => {
+  const [isBookmarked, setIsBookmarked] = useState(false)
```
[the remark this answers is not in the document]
Comment (nole2701) [reply website/src/pages/documents/document.page.tsx:494]: That is for updating the bookmarkButton on the page to say "Bookmark" or "Un-Bookmark" depending on what the state is.

File: graphql/schema.graphql [hunk 2]
```diff
@@ -965,6 +965,10 @@ type Query {
 	"""
 	Retrieves a full document from its unique identifier.
 	"""
+	documentByUuid(id: UUID!): AnnotatedDoc
+	"""
+	Retrieves a full document from its unique identifier.
+	"""
```
[the remark this answers is not in the document]
Comment (nole2701) [reply graphql/schema.graphql:971]: I feel like having it distinctively as UUID makes it clear what type needs to go in.

**Review by CharlieMcVicker — COMMENTED** [review]
You should use the `bookmarkedAt` to decide if a document has been bookmarked. You currently use two different ways to check, which should both be replaced by a check against `bookmarkedAt`.

1. In one place, you set up a piece of state for this. Since this data is already stored in the database and retrieved over GraphQL, we don't have to (and we shouldn't) add a piece of state for it. That creates the possibility that our local state could be out of sync with what is in the database, and that wastes developer hours to debug. Even if you are sure your code does the right thing 100% of the time (no code does), it adds overhead for future developers to have to think about two copies of the same data - is this document bookmarked? - one in the database, another one in this local useState
2. When choosing which mutation to do, you use `Array.some` to check if we fetched this document as a `bookmarkedDocument`. This will get slower and slower as a user bookmarks more documents. Also, it means we need to fetch all the documents they've bookmarked to find out if they bookmarked this one. That also gets slower as we scale.

I think you fix both of these by just passing an `isBookmarked` variable into you `BookmarkButton`. The value of that will be `p.doc.bookmarkedOn.formattedDate !== null`. If your `doc` here doesn't have `bookmarkedOn` getting fetched, then add it to the query. If you end up having to add it to a bunch of queries, think about making a fragment with the common pieces between those queries.

If you don't want to perform that `document.bookmarkedAt.formattedDate !== null` check in a bunch of places, you could also pass a small `fragment` on the `AnnotatedDoc` type into your `BookmarkButton`, and then do that check within your component.

**Review by GracefulLemming — APPROVED** [review]

**Review by CharlieMcVicker — COMMENTED** [review]
LGTM! Go ahead and squash merge.

**Pull request merged**